capsules/src/kv_store.rs

```diff
@@ -0,0 +1,626 @@
+                        if let Err((key, e)) = self.mux_kv.kv.invalidate_key(hashed_key) {
+                            self.hashed_key.replace(key);
+                            self.unhashed_key.take().map(|unhashed_key| {
+                                self.client.map(move |cb| {
+                                    cb.delete_complete(e, unhashed_key);
+                                });
+                            });
+                        }
+                    });
+                } else {
+                    self.unhashed_key.take().map(|unhashed_key| {
+                        self.client.map(move |cb| {
+                            cb.delete_complete(Err(ErrorCode::FAIL), unhashed_key);
+                        });
+                    });
+                }
+            }
+            Operation::Get => {
+                let mut read_allowed = false;
+
+                if result.is_ok() {
+                    let header = KeyHeader::new_from_buf(ret_buf);
+
+                    if header.version == HEADER_VERSION {
+                        let (num, read_ids) = self.valid_ids.get();
+
+                        for i in 0..num {
+                            // If we have permission to read an ID that the data was written with
+                            if read_ids[i] == header.write_id {
+                                read_allowed = true;
+                                break;
+                            }
+                        }
+
+                        if read_allowed {
+                            ret_buf.copy_within(
+                                HEADER_LENGTH..(HEADER_LENGTH + header.length as usize),
+                                0,
+                            );
+                        }
+                    }
+                }
+
+                if !read_allowed {
+                    // Access denied or the header is invalid, zero the buffer
+                    ret_buf.iter_mut().for_each(|m| *m = 0)
+                }
+
+                self.unhashed_key.take().map(|unhashed_key| {
+                    self.client.map(move |cb| {
+                        if read_allowed {
+                            cb.get_complete(result, unhashed_key, ret_buf);
+                        } else {
+                            // The operation failed or the caller doesn't have permission,
+                            // just return an error (and an empty buffer)
+                            cb.get_complete(Err(ErrorCode::FAIL), unhashed_key, ret_buf);
+                        }
+                    });
+                });
+                self.mux_kv.operation.clear();
+            }
+        });
+
+        self.mux_kv.do_next_op();
+    }
+
+    fn invalidate_key_complete(&self, result: Result<(), ErrorCode>, key: &'static mut T) {
+        self.hashed_key.replace(key);
+
+        self.mux_kv.operation.map(|op| match op {
+            Operation::Set | Operation::Get => {
+                unreachable!();
```

Comment:
I advocate for just using a debug or nothing in these sort of instances, rather than embedding a panicking path in the binary which takes more size than a debug!() call and would crash the system if it was ever hit.